components/webext-storage/src/sync/mod.rs

```diff
@@ -0,0 +1,156 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+mod incoming;
+mod outgoing;
+
+#[cfg(test)]
+mod sync_tests;
+
+use serde_derive::*;
+use serde_json;
+use sync15_traits::ServerTimestamp;
+use sync_guid::Guid as SyncGuid;
+
+use incoming::IncomingAction;
+
+type JsonMap = serde_json::Map<String, serde_json::Value>;
+
+/// For use with `#[serde(skip_serializing_if = )]`
+#[inline]
+pub fn is_default<T: PartialEq + Default>(v: &T) -> bool {
+    *v == T::default()
+}
+
+// XXX - need to work out how to consolidate "payload" vs "bso", particularly
+// the timestamp, etc. For now, we represent this in ServerPayload, even though
+// that's never actually stored in this way.
+// XXX - this isn't going to work without other changes.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ServerPayload {
+    guid: SyncGuid,
+    ext_id: String,
+    #[serde(default, skip_serializing_if = "is_default")]
+    data: Option<String>,
+    #[serde(default, skip_serializing_if = "is_default")]
+    deleted: bool,
```

Comment:
What does it mean if `deleted = false` but `data = Nothing`? Is this the same as an empty object?

Reply:
Yeah, it is representing the actual format on the server. I'm still not 100% happy that's the right choice for the server either though - how is this redundant and potentially ambiguous  option better than an explicit data=null?